cmd/juju/model/destroy.go

```diff
@@ -157,18 +164,46 @@ func (c *destroyCommand) Run(ctx *cmd.Context) error {
 	defer api.Close()
 
 	// Attempt to destroy the model.
+	ctx.Infof("Destroying model")
 	err = api.DestroyModel(names.NewModelTag(modelDetails.ModelUUID))
 	if err != nil {
 		return c.handleError(errors.Annotate(err, "cannot destroy model"), modelName)
 	}
 
+	// Wait for model to be destroyed.
+	modelFound := newTimedModelFound(ctx, api, names.NewModelTag(modelDetails.ModelUUID), c.clock)
+	found := modelFound(0)
+	for found {
+		ctx.Infof("Waiting for model to be removed...")
+		found = modelFound(2 * time.Second)
+	}
+
 	err = store.RemoveModel(controllerName, modelName)
 	if err != nil && !errors.IsNotFound(err) {
 		return errors.Trace(err)
 	}
 	return nil
 }
 
+// newTimedModelFound returns a function which waits a given period of time
+// before querying the API server for the status of a model.
+func newTimedModelFound(ctx *cmd.Context, api DestroyModelAPI, tag names.ModelTag, clock clock.Clock) func(time.Duration) bool {
+	return func(wait time.Duration) bool {
+		if wait > 0 {
+			<-clock.After(wait)
+		}
+		res, err := api.ModelInfo([]names.ModelTag{tag})
+		if err != nil {
+			ctx.Infof("Unable to get the model status from the API: %v.", err)
+			return false
+		}
+		if res[0].Error == nil {
+			return true
+		}
+		return params.ErrCode(res[0].Error) != params.CodeUnauthorized
```

Comment:
the intent would be a bit clearer if you used params.IsCodeNotFoundOrCodeUnauthorized